base/ca/src/main/java/com/netscape/ca/CAService.java

```diff
@@ -1716,21 +1706,35 @@ private boolean compareChallengePassword(CertRecord record, String pwd)
             throw new EBaseException(CMS.getUserMessage("CMS_BASE_INVALID_ATTRIBUTE", "metaInfo"));
         }
 
-        String hashpwd = hashPassword(pwd);
-
         // got metaInfo
         String challengeString =
                 (String) metaInfo.get(CertRecord.META_CHALLENGE_PHRASE);
+        String hashpwd = hashPassword(pwd, challengeString);
 
         return challengeString.equals(hashpwd);
     }
 
-    private String hashPassword(String pwd) {
+    private String hashPassword(String pwd, String challengeString) {
+        MessageDigest md = null;
+        String digestAlg = null;
+        try {
+            // Attempt to get algorithm from challengeString to ensure a match
+            digestAlg = challengeString.split("\\{")[1].split("\\}")[0].toUpperCase();
+            digestAlg = (digestAlg.equals("SHA1") || digestAlg.equals("SHA-1")) ? "SHA" : digestAlg;
+        } catch (IndexOutOfBoundsException e) {
+            // Fall back to the default hashing algorithm
+            digestAlg = CryptoUtil.getDefaultHashAlgName();
+        }
+        try {
+            md  = MessageDigest.getInstance(digestAlg);
+        } catch (NoSuchAlgorithmException e) {
+            logger.warn(CMS.getLogMessage("OPERATION_ERROR", e.toString()), e);
+        }
         String salt = "lala123";
-        byte[] pwdDigest = mSHADigest.digest((salt + pwd).getBytes());
+        byte[] pwdDigest = md.digest((salt + pwd).getBytes());
         String b64E = Utils.base64encode(pwdDigest, true);
-
-        return "{SHA}" + b64E;
+        logger.debug("Password hashed with {}", md.getAlgorithm());
+        return String.format("{%s}%s", md.getAlgorithm(), b64E);
     }
```

Comment:
This new change is actually what I was hoping you'd do originally.  However, please watch out for what I pointed out at the other comment, if hashPassword() is used for comparison, this code makes sense, as the challenge is expected to contain the hashing alg;  However, if the hashPassword() is used for the purpose of getting the hash, and storing the challenge for later comparison, then you need to use the default.